### PR TITLE
I2C: GPIO: Convert bit banging I2C controller to use devicetree binding

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Convert the GPIO bit banging I2C driver to configure its instances from devicetree.

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
@galak, @mbolivar-nordic Any idea why the following devicetree snippet utilizing the binding introduced in this PR causes `scripts/dts/gen_legacy_defines.py` to fail?

```
	gpio_i2c0: gpio_i2c {
		status = "okay";
		compatible = "gpio-i2c";
		scl-gpios = <&gpioe 24 0>;
		sda-gpios = <&gpioe 25 0>;
		clock-frequency = <I2C_BITRATE_STANDARD>;
		label = "GPIO_I2C_0";
		#address-cells = <1>;
		#size-cells = <0>;

		fxos8700@1d {
			compatible = "nxp,fxos8700";
			reg = <0x1d>;
			label = "FXOS8700";
			int1-gpios = <&gpioc 6 GPIO_ACTIVE_LOW>;
			int2-gpios = <&gpioc 13 GPIO_ACTIVE_LOW>;
		};
	};
```

The build error is:

```
$ cmake -DBOARD=frdm_k64f ..
Including boilerplate (Zephyr base): /home/brix/Projects/zephyrproject/zephyr/cmake/app/boilerplate.cmake
-- Application: /home/brix/Projects/zephyrproject/zephyr/tests/boards/board_shell
-- Zephyr version: 2.2.99 (/home/brix/Projects/zephyrproject/zephyr)
-- Found Python3: /usr/bin/python3.6 (found suitable exact version "3.6.9") found components:  Interpreter 
-- Board: frdm_k64f
-- Found west: /home/brix/.local/bin/west (found suitable version "0.7.2", minimum required is "0.7.1")
-- Found toolchain: zephyr (/opt/zephyr-sdk)
-- Found BOARD.dts: /home/brix/Projects/zephyrproject/zephyr/boards/arm/frdm_k64f/frdm_k64f.dts
-- Generated zephyr.dts: /home/brix/Projects/zephyrproject/zephyr/tests/boards/board_shell/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /home/brix/Projects/zephyrproject/zephyr/tests/boards/board_shell/build/zephyr/include/generated/devicetree_unfixed.h
Traceback (most recent call last):
  File "/home/brix/Projects/zephyrproject/zephyr/scripts/dts/gen_legacy_defines.py", line 822, in <module>
    main()
  File "/home/brix/Projects/zephyrproject/zephyr/scripts/dts/gen_legacy_defines.py", line 51, in main
    write_regs(node)
  File "/home/brix/Projects/zephyrproject/zephyr/scripts/dts/gen_legacy_defines.py", line 193, in write_regs
    write_reg(reg, "BASE_ADDRESS", hex(reg.addr))
  File "/home/brix/Projects/zephyrproject/zephyr/scripts/dts/gen_legacy_defines.py", line 190, in write_reg
    f"{str2ident(reg.name)}_{base_ident}" if reg.name else None)
  File "/home/brix/Projects/zephyrproject/zephyr/scripts/dts/gen_legacy_defines.py", line 695, in out_node
    node_prefix = node_ident(node)
  File "/home/brix/Projects/zephyrproject/zephyr/scripts/dts/gen_legacy_defines.py", line 302, in node_ident
    str2ident(node.bus_node.matching_compat), node.bus_node.unit_addr)
TypeError: unsupported format string passed to NoneType.__format__
CMake Error at /home/brix/Projects/zephyrproject/zephyr/cmake/dts.cmake:241 (message):
  gen_legacy_defines.py failed with return code: 1
Call Stack (most recent call first):
  /home/brix/Projects/zephyrproject/zephyr/cmake/app/boilerplate.cmake:503 (include)
  /home/brix/Projects/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:24 (include)
  /home/brix/Projects/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:35 (include_boilerplate)
  CMakeLists.txt:4 (find_package)


-- Configuring incomplete, errors occurred!

```

**Comment by henrikbrixandersen** [comment]
> @galak, @mbolivar-nordic Any idea why the following devicetree snippet utilizing the binding introduced in this PR causes `scripts/dts/gen_legacy_defines.py` to fail?

I have added a fix for `scripts/dts/gen_legacy_defines.py` to handle bus parent nodes without `regs` properties. Please review.

**Comment by henrikbrixandersen** [comment]
I would like to see this in v2.3.0, if possible. It is part of a stepping stone path for https://github.com/zephyrproject-rtos/zephyr/pull/23442.

File: drivers/i2c/i2c_gpio.c [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
- * This driver implements an I2C interface by driving two GPIO lines under
- * software control. The GPIO pins used must be preconfigured to to a suitable
- * state, i.e. the SDA pin as open-colector/open-drain with a pull-up resistor
- * (possibly as an external component attached to the pin). When the SDA pin
- * is read it must return the state of the physical hardware line, not just the
- * last state written to it for output.
+ * This driver implements an I2C interface by driving two GPIO lines
+ * under software control. The GPIO pins used must be configured
+ * through devicetree and pinmux with suitable flags, i.e. the SDA pin
+ * as open-colector/open-drain with a pull-up resistor (possibly as an
```
Comment (pabigot) [inline drivers/i2c/i2c_gpio.c:17]: I'd rather see major reformatting like this done in a separate commit.  If there's any actual content change here it's not obvious.

File: drivers/i2c/i2c_gpio.c [hunk 2]
```diff
@@ -24,41 +28,51 @@
...
 /* Driver config */
 struct i2c_gpio_config {
-	char *gpio_name;
-	u8_t scl_pin;
-	u8_t sda_pin;
+	const char *scl_gpio_name;
+	const char *sda_gpio_name;
+	gpio_pin_t scl_pin;
+	gpio_pin_t sda_pin;
+	gpio_flags_t scl_flags;
```
Comment (pabigot) [inline drivers/i2c/i2c_gpio.c:44]: These should be `gpio_dt_flags_t`.

**Review by pabigot — CHANGES_REQUESTED** [review]
Change the flags type for the devicetree-sourced value; the rest is commentary.

File: drivers/i2c/i2c_gpio.c (hunk 2 above)
Comment (henrikbrixandersen) [reply drivers/i2c/i2c_gpio.c]: Fixed, thanks.

File: drivers/i2c/i2c_gpio.c (hunk 1 above)
Comment (henrikbrixandersen) [reply drivers/i2c/i2c_gpio.c]: @pabigot Good point. I have split the reformatting from the content change in two, separate commits.

File: drivers/i2c/i2c_gpio.c [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
+
 /**
  * @file
  * @brief Driver for software driven I2C using GPIO lines
  *
  * This driver implements an I2C interface by driving two GPIO lines under
- * software control. The GPIO pins used must be preconfigured to to a suitable
- * state, i.e. the SDA pin as open-colector/open-drain with a pull-up resistor
+ * software control. The GPIO pins used must be configured (through devicetree and pinmux) with suitable
+ * flags, i.e. the SDA pin as open-colector/open-drain with a pull-up resistor
```
Comment (pabigot) [inline drivers/i2c/i2c_gpio.c]: s/colector/collector/

File: drivers/i2c/i2c_gpio.c [hunk 4]
```diff
@@ -24,41 +27,51 @@
...
-	struct device *gpio;		/* GPIO used for I2C lines */
-	u8_t scl_pin;			/* Pin on gpio used for SCL line */
-	u8_t sda_pin;			/* Pin on gpio used for SDA line */
+	struct device *scl_gpio;	/* GPIO used for I2C SCL line */
+	struct device *sda_gpio;	/* GPIO used for I2C SDA line */
+	gpio_pin_t scl_pin;		/* Pin on gpio used for SCL line */
+	gpio_pin_t sda_pin;		/* Pin on gpio used for SDA line */
 };
 
 static void i2c_gpio_set_scl(void *io_context, int state)
```
Comment (pabigot) [inline drivers/i2c/i2c_gpio.c:60]: Not in scope for this revision, but it seems odd to be passing this as a `void *` when the implementation requires that it be a `struct i2c_gpio_context`.  If that could be cleaned up maybe we could avoid duplicating the pin storage in two structures.

**Review by pabigot — COMMENTED** [review]
One tweak to documentation then I think this is fine.

File: drivers/i2c/i2c_gpio.c (hunk 3 above)
Comment (henrikbrixandersen) [reply drivers/i2c/i2c_gpio.c]: Fixed.

File: drivers/i2c/i2c_gpio.c (hunk 4 above)
Comment (henrikbrixandersen) [reply drivers/i2c/i2c_gpio.c:60]: Yes, agreed. I plan to rework that in a separate PR.

**Review by pabigot — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
Looks very nice.

**Review by MaureenHelm — APPROVED** [review]

**Review by galak — APPROVED** [review]
